compiler/rustc_mir/src/borrow_check/diagnostics/move_errors.rs

```diff
@@ -411,8 +411,11 @@ impl<'a, 'tcx> MirBorrowckCtxt<'a, 'tcx> {
             let suggest = match self.infcx.tcx.get_diagnostic_item(sym::IntoIterator) {
                 Some(def_id) => type_known_to_meet_bound_modulo_regions(
                     &self.infcx,
-                    self.param_env,
-                    self.infcx.tcx.mk_imm_ref(self.infcx.tcx.lifetimes.re_erased, ty),
+                    rustc_middle::ty::ParamEnv::empty(),
```

Comment:
`ParamEnv::empty()` is almost always the wrong thing.

The reason we are getting the ICE "region constraints already solved" is that the `infercx` has already reported all lifetime errors and thus destroyed its region inference system.

Usually when we want to do isolated checks like this we generate a new inference context, so you could wrap the entire function call in a `self.infcx.tcx.infer_ctxt().enter(|infcx| ..)` which will give you a fresh inference context.

The advantage of using a new context and the full `ParamEnv` over your current change is that it will also work in generic contexts where we only know that `&T: IntoIterator` because of where bounds instead of only working for types for which this is known even without bounds.

You may have to use a new lifetime (not `re_erased`, but `infcx.next_region_var(RegionVariableOrigin::Misc(DUMMY_SP))`)

Reply:
I'd tried with `infcx_ctxt().enter`, but until I called `erase_regions` on the `ty`, nothing changed. Looking at this again, it is likely that _just_ calling `erase_regions` might have been enough...

Reply:
Updated after our 1:1 chat